spec/integration/util/windows/security_spec.rb

```diff
@@ -553,6 +572,100 @@ def grant_everyone_full_access(path)
           # end
         end
 
+        describe "#mode=" do
```

Comment:
Why do we have `describe "#mode=" do` a second time?

Reply:
Same name, different context

```
Puppet::Util::Windows::Security
  file
    behaves like a securable object
      on a volume that doesn't support ACLs
        should return nil owner
        should return nil group
        should return nil mode
      on a volume that supports ACLs
        for a normal user
          should preserve inherited full control for SYSTEM when setting owner and group
          #supports_acl?
            should accept c:/
            should accept c:\
            should accept c:/windows/system32
            should accept \\localhost\C$
            should accept \\127.0.0.1\C$\foo
            should raise an exception if it cannot get volume information
          #owner=
            should allow setting to the current user
            should raise an exception when setting to a different user
          #owner
            it should not be empty
            should raise an exception if an invalid path is provided
          #group=
            should allow setting to a group the current owner is a member of
            should allow setting to a group the current owner is not a member of
          #group
            should not be empty
            should raise an exception if an invalid path is provided
          #mode=
            should enforce mode 0
            should enforce mode 100
            should enforce mode 200
            should enforce mode 300
            should enforce mode 400
            should enforce mode 500
            should enforce mode 600
            should enforce mode 700
            should round-trip all 128 modes that do not require deny ACEs, where owner and group are different
            should round-trip all 54 modes that do not require deny ACEs, where owner and group are same
            should preserve full control for SYSTEM when setting mode
            should raise an exception if an invalid path is provided
            for modes that require deny aces
              should map everyone to group and owner
              should combine user and group modes when owner and group sids are equal
            for read-only objects
              should make them writable if any sid has write permission
              should leave them read-only if no sid has write permission and should allow full access for SYSTEM
          #mode
            should report when extra aces are encounted
            should return deny aces
            should skip inherit-only ace
            should raise an exception if an invalid path is provided
          inherited access control entries
            should be absent when the access control list is protected, and should not remove SYSTEM
            should be present when the access control list is unprotected
        for an administrator
          #owner=
            should accept the guest sid
            should accept a user sid
            should accept a group sid
            should raise an exception if an invalid sid is provided
            should raise an exception if an invalid path is provided
          #group=
            should accept the test group
            should accept a group sid
            should accept a user sid
            should combine owner and group rights when they are the same sid
            should raise an exception if an invalid sid is provided
            should raise an exception if an invalid path is provided
          when the sid is NULL
            should retrieve an empty owner sid (PENDING: Not yet implemented)
            should retrieve an empty group sid (PENDING: Not yet implemented)
          when the sid refers to a deleted trustee
            should retrieve the user sid
            should retrieve the group sid
          #mode
            should deny all access when the DACL is empty, including SYSTEM
          #mode=
            should round-trip all 128 modes that do not require deny ACEs, where owner and group are different
            should round-trip all 54 modes that do not require deny ACEs, where owner and group are same
            should round-trip all 128 modes that do not require deny ACEs, when running as a service as SYSTEM
            should round-trip all 54 modes that do not require deny ACEs, when running as scheduled task as SYSTEM
          when the parent directory
            is writable and executable
              and sticky bit is set
                should allow child owner
                should allow parent owner
                should deny group
                should deny other
              and sticky bit is not set
                should allow child owner
                should allow parent owner
                should allow group
                should allow other
            is not writable
              behaves like only child owner
                should allow child owner
                should deny parent owner
                should deny group
                should deny other
            is not executable
              behaves like only child owner
                should allow child owner
                should deny parent owner
                should deny group
                should deny other
    locked files
      should get the owner
      should get the group
      should get the mode
  directory
    behaves like a securable object
      on a volume that doesn't support ACLs
        should return nil owner
        should return nil group
        should return nil mode
      on a volume that supports ACLs
        for a normal user
          should preserve inherited full control for SYSTEM when setting owner and group
          #supports_acl?
            should accept c:/
            should accept c:\
            should accept c:/windows/system32
            should accept \\localhost\C$
            should accept \\127.0.0.1\C$\foo
            should raise an exception if it cannot get volume information
          #owner=
            should allow setting to the current user
            should raise an exception when setting to a different user
          #owner
            it should not be empty
            should raise an exception if an invalid path is provided
          #group=
            should allow setting to a group the current owner is a member of
            should allow setting to a group the current owner is not a member of
          #group
            should not be empty
            should raise an exception if an invalid path is provided
          #mode=
            should enforce mode 0
            should enforce mode 100
            should enforce mode 200
            should enforce mode 300
            should enforce mode 400
            should enforce mode 500
            should enforce mode 600
            should enforce mode 700
            should round-trip all 128 modes that do not require deny ACEs, where owner and group are different
            should round-trip all 54 modes that do not require deny ACEs, where owner and group are same
            should preserve full control for SYSTEM when setting mode
            should raise an exception if an invalid path is provided
            for modes that require deny aces
              should map everyone to group and owner
              should combine user and group modes when owner and group sids are equal
            for read-only objects
              should make them writable if any sid has write permission
              should leave them read-only if no sid has write permission and should allow full access for SYSTEM
          #mode
            should report when extra aces are encounted
            should return deny aces
            should skip inherit-only ace
            should raise an exception if an invalid path is provided
          inherited access control entries
            should be absent when the access control list is protected, and should not remove SYSTEM
            should be present when the access control list is unprotected
        for an administrator
          #owner=
            should accept the guest sid
            should accept a user sid
            should accept a group sid
            should raise an exception if an invalid sid is provided
            should raise an exception if an invalid path is provided
          #group=
            should accept the test group
            should accept a group sid
            should accept a user sid
            should combine owner and group rights when they are the same sid
            should raise an exception if an invalid sid is provided
            should raise an exception if an invalid path is provided
          when the sid is NULL
            should retrieve an empty owner sid (PENDING: Not yet implemented)
            should retrieve an empty group sid (PENDING: Not yet implemented)
          when the sid refers to a deleted trustee
            should retrieve the user sid
            should retrieve the group sid
          #mode
            should deny all access when the DACL is empty, including SYSTEM
          #mode=
            should round-trip all 128 modes that do not require deny ACEs, where owner and group are different
            should round-trip all 54 modes that do not require deny ACEs, where owner and group are same
            should round-trip all 128 modes that do not require deny ACEs, when running as a service as SYSTEM
            should round-trip all 54 modes that do not require deny ACEs, when running as scheduled task as SYSTEM
          when the parent directory
            is writable and executable
              and sticky bit is set
                should allow child owner
                should allow parent owner
                should deny group
                should deny other
              and sticky bit is not set
                should allow child owner
                should allow parent owner
                should allow group
                should allow other
            is not writable
              behaves like only child owner
                should allow child owner
                should deny parent owner
                should deny group
                should deny other
            is not executable
              behaves like only child owner
                should allow child owner
                should deny parent owner
                should deny group
                should deny other
    inheritable aces
      should be applied to child objects
  security descriptor
    preserves aces for other users
    changes the sid for all aces that were assigned to the old owner
    preserves INHERIT_ONLY_ACEs
    allows deny ACEs with inheritance
    when managing mode
      removes aces for sids that are neither the owner nor group
```

Reply:
So reducing this down to relevant bits (noting these sets are repeated on `file` and `directory`):

```
Puppet::Util::Windows::Security
  file / directory
    behaves like a securable object
      on a volume that supports ACLs
        for a normal user (Puppet.features.stubs(:root?).returns(false))
          #mode=
            should round-trip all 128 modes that do not require deny ACEs, where owner and group are different
            should round-trip all 54 modes that do not require deny ACEs, where owner and group are same
        for an administrator (guarded on :if => (Puppet.features.root? && Puppet.features.microsoft_windows?)) - *defaults owner to Guest*
          #mode=
            should round-trip all 128 modes that do not require deny ACEs, where owner and group are different
            should round-trip all 54 modes that do not require deny ACEs, where owner and group are same
            should round-trip all 128 modes that do not require deny ACEs, when running as a service as SYSTEM
            should round-trip all 54 modes that do not require deny ACEs, when running as scheduled task as SYSTEM
```

So the initial intent was to document differences specifically when a `root` user (i.e. when rspec is run as Jenkins user or SYSTEM or elevated local admin), which the testing layout doesn't make clear.  But the tests aren't doing that because they're explicitly setting group / owner (instead of using the context of the user). Being explicit is fine / desirable, but note that we can do that regardless of what user we're running as, because of how NTFS perms work. Let me break that down a bit more:

`normal user #mode= should round-trip all 128 modes that do not require deny ACEs, where owner and group are different`
* uses current user / group - does not set explicitly
    * when run under SYSTEM, this is Administrators, SYSTEM
    * when run under local Administrator, this is Administrators, None

`normal user #mode= should round-trip all 54 modes that do not require deny ACEs, where owner and group are same`
* uses current user / explicitly sets group to same as user
    * when run under SYSTEM, this is Administrators, Administrators
    * when run under local Administrator, this is Administrators, Administrators

`for an administrator should round-trip all 128 modes that do not require deny ACEs, where owner and group are different`
* explicitly sets user to `current_user`, group to `Users`
    * when run under SYSTEM, this is SYSTEM, Users
    * when run under local Administrator, this is Administrator, Users

* `for an administrator should round-trip all 54 modes that do not require deny ACEs, where owner and group are same`
* uses current user / explicitly sets group to same as user (picks up before that sets default owner to Guest)
    * when run under SYSTEM, this is Guest, Guest
    * when run under local Administrator, this is Guest, Guest

`for an administrator should round-trip all 128 modes that do not require deny ACEs, when running as a service as SYSTEM`
* explicitly sets owner to Administrators, group to SYSTEM
    * when run under SYSTEM, this is Administrators, SYSTEM
    * when run under local Administrator, this is Administrators, SYSTEM

`for an administrator should round-trip all 54 modes that do not require deny ACEs, when running as scheduled task as SYSTEM`
* explicitly sets owner to SYSTEM, group to SYSTEM
    * when run under SYSTEM, this is SYSTEM, SYSTEM
    * when run under local Administrator, this is SYSTEM, SYSTEM


To summarize the issue that we have here:

* `should round-trip all 128 modes that do not require deny ACEs, where owner and group are different` tests, under `normal user` or `for an administrator` are redundant - they assert the same thing and the new set can be removed
* `should round-trip all 54 modes that do not require deny ACEs, where owner and group are same` tests, under `normal user` or `for an administrator` are redundant - they assert the same thing and the new set can be removed
* `for an administrator should round-trip all 128 modes that do not require deny ACEs, when running as a service as SYSTEM` are net new and valuable - but they can be put in with the original set of `#mode=` tests as they don't require being an admin to run
* `for an administrator should round-trip all 54 modes that do not require deny ACEs, when running as scheduled task as SYSTEM` are net new and valuable - but they can be put in with the original set of `#mode=` tests as they don't require being an admin to run


I'm going to open a modified version of this PR with these changes to run through AppVeyor.

Reply:
Addendum - it looks like `for an administrator should round-trip all 54 modes that do not require deny ACEs, when running as scheduled task as SYSTEM` can only be run when `root?` because setting an owner of `SYSTEM` is only allowed under that circumstance.

Updated PR is at #6942